### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by pbrisbin**
Best context is probably the [Rendered README](https://github.com/freckle/hs-opentelemetry-awsxray/blob/pb/first/README.md).

### CODE REVIEW

File: library/OpenTelemetry/AWSXRay/Propagator.hs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  (Context, insertBaggage, insertSpan, lookupBaggage, lookupSpan)
+import OpenTelemetry.Propagator
+import OpenTelemetry.Trace.Core (getSpanContext, wrapSpanContext)
+
+awsXRayContextPropagator :: Propagator Context RequestHeaders ResponseHeaders
+awsXRayContextPropagator = awsXRayContextPropagatorOnError $ \_ _ -> pure ()
+
+awsXRayContextPropagatorOnError
+  :: (RequestHeaders -> String -> IO ())
+  -- ^ Called on failure to find or parse an @X-Amzn-TraceId@ header
```
Comment (pbrisbin) [inline library/OpenTelemetry/AWSXRay/Propagator.hs]:
```suggestion
  -- ^ Called on failure to find or parse an @X-Amzn-Trace-Id@ header
```

File: library/OpenTelemetry/AWSXRay/TraceInfo.hs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+  (epoch, unique) = BS.splitAt 8 $ traceIdBaseEncodedByteString Base16 traceId
+
+prefix :: String -> Either String a -> Either String a
+prefix p = first (\e -> p <> ": " <> e)
+
+bsToKeyValues :: ByteString -> Either String [(ByteString, ByteString)]
+bsToKeyValues = traverse go . bsSplitOn ';'
+ where
+  go bs = case bsSplitOn '=' bs of
+    k : vs -> Right (k, mconcat vs)
```
Comment (pbrisbin) [inline library/OpenTelemetry/AWSXRay/TraceInfo.hs:107]: Could guard against `BS.null k` here. :shrug:

File: library/OpenTelemetry/Trace/Setup.hs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  bracket (initializeGlobalTracerProvider setup) shutdownTracerProvider
+
+initializeGlobalTracerProvider
+  :: MonadIO m
+  => (TracerProviderOptions -> TracerProviderOptions)
+  -> m TracerProvider
+initializeGlobalTracerProvider setup = liftIO $ do
+  t <- initializeTracerProvider setup
+  setGlobalTracerProvider t
+  pure t
```
Comment (pbrisbin) [inline library/OpenTelemetry/Trace/Setup.hs]:
```suggestion
  t <$ setGlobalTracerProvider t
```

File: library/OpenTelemetry/AWSXRay/Baggage.hs [hunk 4]
```diff
@@ -0,0 +1,52 @@
+-- | Functions for dealing with 'Baggage'
```
Comment (mjgpy3) [inline library/OpenTelemetry/AWSXRay/Baggage.hs:1]: We could all use 

> Functions for dealing with 'Baggage'

in our lives.

File: README.md [hunk 5]
```diff
@@ -1,37 +1,55 @@
...
-## Rename your package
+[sdk]: https://hackage.haskell.org/package/hs-opentelemetry-sdk
+[xray]: https://docs.aws.amazon.com/xray/latest/devguide/xray-api-sendingdata.html#xray-api-traceids
 
-```sh
-sed -i s/haskell-library-template/my-name/ ./**/*
-```
+## Usage
+
+The API currently exposed by the `hs-opentelemetry-sdk` package lacks a
```
Comment (mjgpy3) [inline README.md:20]: I'm sure I'm just missing context on this somewhere but what about upstreaming vs. adding a new package?

File: library/OpenTelemetry/AWSXRay/Baggage.hs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+--
+-- We ported specs about 'Baggage' from the Java Propagator:
+--
+-- <https://github.com/open-telemetry/opentelemetry-java-contrib/blob/04a2a481934be614e73e41194107ffdd767bc507/aws-xray-propagator/src/test/java/io/opentelemetry/contrib/awsxray/propagator/AwsXrayPropagatorTest.java#L227-L238>
+--
+-- And using the obvious 'encodeBaggageHeader'/'decodeBaggageHeader' functions
```
Comment (mjgpy3) [inline library/OpenTelemetry/AWSXRay/Baggage.hs:7]: That's too bad... Is it worth encapsulating these in our specs somewhere so they're not part of "the API we maintain" :tm: .

File: README.md (hunk 5 above)
Comment (mjgpy3) [reply README.md:20]: Is it just that this is super XRay-specific so there's no natural home for it?

File: library/OpenTelemetry/AWSXRay/TraceInfo.hs [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+fromXRayHeader :: ByteString -> Either String TraceInfo
+fromXRayHeader bs = do
+  kv <- bsToKeyValues bs
+
+  root <- note "Root not present" $ lookup "Root" kv
+  traceId <- case bsSplitOn '-' root of
+    ["1", epoch, unique] -> do
+      -- epoch can be "at most" 8 hex and we should zero-pad the
+      -- short values. NB. This relies on hex being 1 char per char
```
Comment (mjgpy3) [inline library/OpenTelemetry/AWSXRay/TraceInfo.hs]: :thinking:

Dumb question warning: why would it not be "1 char per char"?

File: library/OpenTelemetry/AWSXRay/TraceInfo.hs (hunk 2 above)
Comment (mjgpy3) [reply library/OpenTelemetry/AWSXRay/TraceInfo.hs:107]: If it makes an "odd gotcha" for consumers into a specific error I'm for it.

Though I guess where do you draw the line? Check it's not only whitespace? Check it's not weird characters that xray doesn't support?...

File: library/OpenTelemetry/Trace/Setup.hs [hunk 8]
```diff
@@ -0,0 +1,45 @@
+-- | Initialization of a 'TracerProvider' with modified 'TracerProviderOptions'
+module OpenTelemetry.Trace.Setup
```
Comment (mjgpy3) [inline library/OpenTelemetry/Trace/Setup.hs:2]: Re-implementation came out very clean, nice job :+1:

File: package.yaml [hunk 9]
```diff
@@ -24,6 +23,9 @@ ghc-options:
   - -Wno-safe
 
 when:
+  - condition: "impl(ghc >= 9.2)"
```
Comment (mjgpy3) [inline package.yaml:26]: Should this be a template change too?

File: tests/OpenTelemetry/AWSXRay/TraceInfoSpec.hs [hunk 10]
```diff
@@ -0,0 +1,90 @@
+module OpenTelemetry.AWSXRay.TraceInfoSpec
```
Comment (mjgpy3) [inline tests/OpenTelemetry/AWSXRay/TraceInfoSpec.hs:1]: Are these the specs we ported? Should we link somewhere here too just for coherency/consistency?

**Review by mjgpy3 — APPROVED** [review]
Looks good to me. I left some thoughts but you can take 'em or leave 'em.

File: README.md (hunk 5 above)
Comment (pbrisbin) [reply README.md:20]: Well, I guess what I meant to say was more like

> Until we go fix it, the API exposed by the...

File: library/OpenTelemetry/AWSXRay/TraceInfo.hs (hunk 7 above)
Comment (pbrisbin) [reply library/OpenTelemetry/AWSXRay/TraceInfo.hs]: I guess it's just that it relies on it being _hex_. A `ByteString` in some other base or encoding would behave differently when you do a `Char`-based operation, I think. I'd _guess_ that `'0'` is one of those characters this is going to work no matter the base or encoding, but I'm not sure! I can definitely nuke the comment if it's more confusing than helpful.

File: package.yaml (hunk 9 above)
Comment (pbrisbin) [reply package.yaml:26]: Yeah, I made a few updates to the template after using it and finding issues. This fix was included in [the PR](https://github.com/freckle/haskell-library-template/pull/8).

File: tests/OpenTelemetry/AWSXRay/TraceInfoSpec.hs (hunk 10 above)
Comment (pbrisbin) [reply tests/OpenTelemetry/AWSXRay/TraceInfoSpec.hs:1]: No, I wrote them fresh. The only specs I "ported" was the `Baggage` de/encoding (and they weren't really ported so much as just written to represent similar cases), and the comment about that includes a link.

File: library/OpenTelemetry/AWSXRay/Baggage.hs (hunk 6 above)
Comment (pbrisbin) [reply library/OpenTelemetry/AWSXRay/Baggage.hs:7]: I'm not sure what you mean, can you rephrase?

File: library/OpenTelemetry/AWSXRay/TraceInfo.hs (hunk 2 above)
Comment (pbrisbin) [reply library/OpenTelemetry/AWSXRay/TraceInfo.hs:107]: > Though I guess where do you draw the line?

That's kind of where I'm at.

Also though, we have no code that generates `Baggage` values, nor code that uses them. We're simply passing them along from one place to another. I'd imagine an empty key is unexpected and invalid, but it could be an overreach for us to drop it on the floor when we can easily just pass it along.

File: library/OpenTelemetry/AWSXRay/Baggage.hs (hunk 6 above)
Comment (mjgpy3) [reply library/OpenTelemetry/AWSXRay/Baggage.hs:7]: So right now these would be exposed by our package but I read these comments as saying that they're for testing.

So I'm suggesting not making them part of our exposed interface since there's typically more potential maintenance burden/headache, the more you expose.

File: package.yaml (hunk 9 above)
Comment (mjgpy3) [reply package.yaml:26]: Thank you!

File: tests/OpenTelemetry/AWSXRay/TraceInfoSpec.hs (hunk 10 above)
Comment (mjgpy3) [reply tests/OpenTelemetry/AWSXRay/TraceInfoSpec.hs:1]: Gotcha, makes sense!

File: library/OpenTelemetry/AWSXRay/TraceInfo.hs (hunk 2 above)
Comment (mjgpy3) [reply library/OpenTelemetry/AWSXRay/TraceInfo.hs:107]: Yep I agree. I used to do a little open-source on the aws/azure terraform providers...

They would often break because of little things like AWS increasing the number of characters allowed in a field (the providers duplicated the "rules", and the rules changed, and consumers of the providers only saw the AWS docs).

File: library/OpenTelemetry/AWSXRay/Baggage.hs (hunk 6 above)
Comment (pbrisbin) [reply library/OpenTelemetry/AWSXRay/Baggage.hs:7]: >  I read these comments as saying that they're for testing.

Hmm, no, they're used by the real thing to de/encode `Baggage`.

That said, they are _internal_. But I typically just don't try to slice up my package with public/private lines. I know some folks like to make a `.Internal`, or even maintain `modules` vs `extra-modules`, but I've never found that worthwhile. If I were going to go down that route, more than just these functions could be private. Only the `IdGenerator` and `Propogator` (and the `Setup` stuff) are actually necessary public API.

Comment (mjgpy3) [reply library/OpenTelemetry/AWSXRay/Baggage.hs:7]: Got it.

File: README.md (hunk 5 above)
Comment (pbrisbin) [reply README.md:20]: Added some language near this to indicate we should (and plan to) upstream something to make this unnecessary.
